ffi/jpeg.lua

```diff
@@ -0,0 +1,99 @@
+--[[--
+Module for JPEG decoding/encoding.
+
+This is a LuaJIT FFI wrapper for turbojpeg.
+
+@module ffi.jpeg
+]]
+
+local BB = require("ffi/blitbuffer")
+local ffi = require("ffi")
+local _ = require("ffi/turbojpeg_h")
+
+local turbojpeg
+if ffi.os == "Windows" then
+    turbojpeg = ffi.load("libs/libturbojpeg.dll")
+elseif ffi.os == "OSX" then
+    turbojpeg = ffi.load("libs/libturbojpeg.dylib")
+else
+    turbojpeg = ffi.load("libs/libturbojpeg.so")
+end
+
+local Jpeg = {}
+
+function Jpeg.openDocument(filename, color)
+    local fh = io.open(filename, "rb")
+    assert(fh, "couldn't open JPG file")
+    local data = fh:read("*a")
+    fh:close()
+
+    return Jpeg.openDocumentFromMem(data, color)
+end
+
+function Jpeg.openDocumentFromMem(data, color)
+    local handle = turbojpeg.tjInitDecompress()
+    assert(handle, "no TurboJPEG API decompressor handle")
+
+    local width = ffi.new("int[1]")
+    local height = ffi.new("int[1]")
+    local jpegsubsamp = ffi.new("int[1]")
+
+    turbojpeg.tjDecompressHeader2(handle, ffi.cast("unsigned char*", data), #data, width, height, jpegsubsamp)
+    assert(width[0] > 0 and height[0] > 0, "image dimensions")
+
+    local image_bb
+    local format
+    local components
+    if color then
+        image_bb = BB.new(width[0], height[0], BB.TYPE_BBRGB24)
+        components = 3
+        format = turbojpeg.TJPF_RGB
+    else
+        image_bb = BB.new(width[0], height[0], BB.TYPE_BB8)
+        components = 1
+        format = turbojpeg.TJPF_GRAY
+    end
+
+    if turbojpeg.tjDecompress2(handle, ffi.cast("unsigned char*", data), #data,
```

Comment:
This one takes a `const unsigned char*`. Speaking of, I'd possibly switch to `tjDecompressHeader3`, which *also* takes a const (as it should), to avoid the puzzling madness.

Reply:
I have to look into what I should do with the colorspace tomorrow.

Reply:
Assuming you meant the subsampling constants, there's a `TJSAMP_GRAY` for the `TJPF_GRAY` pixel format ;).